editor/src/components/editor/persistence/persistence.spec.ts

```diff
@@ -0,0 +1,528 @@
+  },
+  setItem: async (id: string, project: LocalProject<PersistentModel>) => {
+    localProjects[id] = project
+  },
+  removeItem: async (id: string) => {
+    delete localProjects[id]
+  },
+  keys: async () => {
+    return Object.keys(localProjects)
+  },
+}))
+
+function setupTest(saveThrottle: number = 0) {
+  let capturedData = {
+    newProjectId: undefined as string | undefined,
+    updatedFiles: {} as { [fileName: string]: AssetFile },
+    dispatchedActions: [] as Array<EditorAction>,
+    projectNotFound: false,
+    createdOrLoadedProject: undefined as PersistentModel | undefined,
+  }
+  const testDispatch: EditorDispatch = (actions: ReadonlyArray<EditorAction>) => {
+    capturedData.dispatchedActions.push(...actions)
+    fastForEach(actions, (action) => {
+      if (action.action === 'SET_PROJECT_ID') {
+        capturedData.newProjectId = action.id
+      } else if (action.action === 'UPDATE_FILE' && isAssetFile(action.file)) {
+        capturedData.updatedFiles[action.filePath] = action.file
+      }
+    })
+  }
+  const onProjectNotFound = () => (capturedData.projectNotFound = true)
+  const onCreatedOrLoadedProject = (
+    _projectId: string,
+    _projectName: string,
+    createdOrLoadedProject: PersistentModel,
+  ) => (capturedData.createdOrLoadedProject = createdOrLoadedProject)
+
+  const testMachine = new PersistenceMachine(
+    PersistenceBackend,
+    testDispatch,
+    onProjectNotFound,
+    onCreatedOrLoadedProject,
+    saveThrottle,
+  )
+  return {
+    capturedData: capturedData,
+    testDispatch: testDispatch,
+    testMachine: testMachine,
+  }
+}
+
+describe('Saving', () => {
+  it('Saves locally when logged out', async () => {
+    const { capturedData, testMachine } = setupTest()
+
+    testMachine.createNew(ProjectName, BaseModel)
+    await delay(20)
+    testMachine.save(ProjectName, FirstRevision, 'force')
+    await delay(20)
+    testMachine.stop()
+
+    expect(capturedData.newProjectId).toBeDefined()
+    expect(mockSaveLog[capturedData.newProjectId!]).toBeUndefined()
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]).toBeDefined()
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]!.model).toEqual(FirstRevision)
+  })
+
+  it('Saves to server when logged in', async () => {
+    const { capturedData, testMachine } = setupTest()
+
+    testMachine.login()
+    await delay(20)
+    testMachine.createNew(ProjectName, BaseModel)
+    await delay(20)
+    testMachine.save(ProjectName, FirstRevision, 'force')
+    await delay(20)
+    testMachine.stop()
+
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]).toBeUndefined()
+    expect(capturedData.newProjectId).toBeDefined()
+    expect(mockSaveLog[capturedData.newProjectId!]).toEqual([BaseModel, FirstRevision])
+  })
+
+  it('Throttles saves', async () => {
+    const { capturedData, testMachine } = setupTest(10000)
+
+    testMachine.login()
+    await delay(20)
```

Comment:
we should use Jest's mock timers because this file adds at least 500ms of extra delays to the test run!

Reply:
wait, I'm realizing you don't have the delays becuase of throttling – you have the delays to make sure that the state machine completed running? :O that's kinda horrifying!

Reply:
Yeah. This is because you can't really fire an event and track when _that_ event has been completely processed (or even handled). I thought about trying to give all events an ID and then tracking which events were handled in a given run, which I think we could do in the interpreter, but that doesn't necessarily correspond to e.g. a specific save actually completing (since the event could result in it being handled on the next cycle)